// @aws-cdk/aws-sqs/lib/sqs-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQSMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class SQSMetrics {
    static numberOfMessagesSentAverage(dimensions) {
        return {
            namespace: 'AWS/SQS',
            metricName: 'NumberOfMessagesSent',
            dimensions,
            statistic: 'Average',
        };
    }
    static approximateNumberOfMessagesDelayedAverage(dimensions) {
        return {
            namespace: 'AWS/SQS',
            metricName: 'ApproximateNumberOfMessagesDelayed',
            dimensions,
            statistic: 'Average',
        };
    }
    static numberOfMessagesReceivedAverage(dimensions) {
        return {
            namespace: 'AWS/SQS',
            metricName: 'NumberOfMessagesReceived',
            dimensions,
            statistic: 'Average',
        };
    }
    static numberOfMessagesDeletedAverage(dimensions) {
        return {
            namespace: 'AWS/SQS',
            metricName: 'NumberOfMessagesDeleted',
            dimensions,
            statistic: 'Average',
        };
    }
    static approximateNumberOfMessagesNotVisibleAverage(dimensions) {
        return {
            namespace: 'AWS/SQS',
            metricName: 'ApproximateNumberOfMessagesNotVisible',
            dimensions,
            statistic: 'Average',
        };
    }
    static approximateNumberOfMessagesVisibleAverage(dimensions) {
        return {
            namespace: 'AWS/SQS',
            metricName: 'ApproximateNumberOfMessagesVisible',
            dimensions,
            statistic: 'Average',
        };
    }
    static approximateAgeOfOldestMessageAverage(dimensions) {
        return {
            namespace: 'AWS/SQS',
            metricName: 'ApproximateAgeOfOldestMessage',
            dimensions,
            statistic: 'Average',
        };
    }
    static numberOfEmptyReceivesAverage(dimensions) {
        return {
            namespace: 'AWS/SQS',
            metricName: 'NumberOfEmptyReceives',
            dimensions,
            statistic: 'Average',
        };
    }
    static sentMessageSizeAverage(dimensions) {
        return {
            namespace: 'AWS/SQS',
            metricName: 'SentMessageSize',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.SQSMetrics = SQSMetrics;
