// @aws-amplify/ui-components/dist/collection/components/amplify-totp-setup/amplify-totp-setup.js
import { Auth } from '@aws-amplify/auth';
import { I18n, Logger } from '@aws-amplify/core';
import { Component, Prop, State, h, Host } from '@stencil/core';
import QRCode from 'qrcode';
import { MfaOption } from '../../common/types/auth-types';
import { Translations } from '../../common/Translations';
import { NO_AUTH_MODULE_FOUND, SETUP_TOTP, SUCCESS } from '../../common/constants';
import { dispatchToastHubEvent, dispatchAuthStateChangeEvent } from '../../common/helpers';
import { checkContact } from '../../common/auth-helpers';
const logger = new Logger('TOTP');
export class AmplifyTOTPSetup {
    constructor() {
        this.inputProps = {
            autoFocus: true,
        };
        /** Auth state change handler for this component */
        this.handleAuthStateChange = dispatchAuthStateChangeEvent;
        /** Used for header text in totp setup component */
        this.headerText = Translations.TOTP_HEADER_TEXT;
        /** Used for customizing the issuer string in the qr code image */
        this.issuer = Translations.TOTP_ISSUER;
        this.code = null;
        this.setupMessage = null;
        this.qrCodeInput = null;
        this.loading = false;
    }
    componentWillLoad() {
        this.setup();
    }
    buildOtpAuthPath(user, issuer, secretKey) {
        return `otpauth://totp/${issuer}:${user.username}?secret=${secretKey}&issuer=${issuer}`;
    }
    async onTOTPEvent(event, data, user) {
        logger.debug('on totp event', event, data);
        if (event === SETUP_TOTP && data === SUCCESS) {
            await checkContact(user, this.handleAuthStateChange);
        }
    }
    handleTotpInputChange(event) {
        this.setupMessage = null;
        this.qrCodeInput = event.target.value;
    }
    async generateQRCode(codeFromTotp) {
        try {
            this.qrCodeImageSource = await QRCode.toDataURL(codeFromTotp);
        }
        catch (error) {
            dispatchToastHubEvent(error);
        }
    }
    async setup() {
        this.setupMessage = null;
        const encodedIssuer = encodeURI(I18n.get(this.issuer));
        if (!Auth || typeof Auth.setupTOTP !== 'function') {
            throw new Error(NO_AUTH_MODULE_FOUND);
        }
        this.loading = true;
        try {
            const secretKey = await Auth.setupTOTP(this.user);
            logger.debug('secret key', secretKey);
            this.code = this.buildOtpAuthPath(this.user, encodedIssuer, secretKey);
            this.generateQRCode(this.code);
        }
        catch (error) {
            dispatchToastHubEvent(error);
            logger.debug(I18n.get(Translations.TOTP_SETUP_FAILURE), error);
        }
        finally {
            this.loading = false;
        }
    }
    async verifyTotpToken(event) {
        if (event) {
            event.preventDefault();
        }
        if (!this.qrCodeInput) {
            logger.debug('No TOTP Code provided');
            return;
        }
        const user = this.user;
        if (!Auth || typeof Auth.verifyTotpToken !== 'function' || typeof Auth.setPreferredMFA !== 'function') {
            throw new Error(NO_AUTH_MODULE_FOUND);
        }
        try {
            await Auth.verifyTotpToken(user, this.qrCodeInput);
            await Auth.setPreferredMFA(user, MfaOption.TOTP);
            this.setupMessage = I18n.get(Translations.TOTP_SUCCESS_MESSAGE);
            logger.debug(I18n.get(Translations.TOTP_SUCCESS_MESSAGE));
            await this.onTOTPEvent(SETUP_TOTP, SUCCESS, user);
        }
        catch (error) {
            this.setupMessage = I18n.get(Translations.TOTP_SETUP_FAILURE);
            logger.error(error);
        }
    }
    render() {
        return (h(Host, null,
            h("amplify-form-section", { headerText: I18n.get(this.headerText), submitButtonText: I18n.get(Translations.TOTP_SUBMIT_BUTTON_TEXT), handleSubmit: event => this.verifyTotpToken(event), loading: this.loading },
                h("div", { class: "totp-setup" },
                    h("img", { src: this.qrCodeImageSource, alt: I18n.get(Translations.QR_CODE_ALT) }),
                    h("amplify-form-field", { label: I18n.get(Translations.TOTP_LABEL), inputProps: this.inputProps, fieldId: "totpCode", name: "totpCode", handleInputChange: event => this.handleTotpInputChange(event) })))));
    }
    static get is() { return "amplify-totp-setup"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() { return {
        "$": ["amplify-totp-setup.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["amplify-totp-setup.css"]
    }; }
    static get properties() { return {
        "user": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "CognitoUserInterface",
                "resolved": "CognitoUserInterface",
                "references": {
                    "CognitoUserInterface": {
                        "location": "import",
                        "path": "../../common/types/auth-types"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Used in order to configure TOTP for a user"
            }
        },
        "handleAuthStateChange": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "AuthStateHandler",
                "resolved": "(nextAuthState: AuthState, data?: object) => void",
                "references": {
                    "AuthStateHandler": {
                        "location": "import",
                        "path": "../../common/types/auth-types"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Auth state change handler for this component"
            },
            "defaultValue": "dispatchAuthStateChangeEvent"
        },
        "headerText": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Used for header text in totp setup component"
            },
            "attribute": "header-text",
            "reflect": false,
            "defaultValue": "Translations.TOTP_HEADER_TEXT"
        },
        "issuer": {
            "type": "string",
            "mutable": false,
            "complexType": {
                "original": "string",
                "resolved": "string",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": "Used for customizing the issuer string in the qr code image"
            },
            "attribute": "issuer",
            "reflect": false,
            "defaultValue": "Translations.TOTP_ISSUER"
        }
    }; }
    static get states() { return {
        "code": {},
        "setupMessage": {},
        "qrCodeImageSource": {},
        "qrCodeInput": {},
        "loading": {}
    }; }
}
