// @aws-amplify/ui-components/dist/cjs/amplify-s3-image.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-1f3b9bdc.js');
const core = require('@aws-amplify/core');
const storageTypes = require('./storage-types-fa93d917.js');
require('./constants-4c621171.js');
require('@aws-amplify/storage');
const storageHelpers = require('./storage-helpers-5d59e8b5.js');

const amplifyS3ImageCss = ":host{height:inherit;width:inherit;--height:inherit;--width:inherit}img{height:var(--height);width:var(--width)}";

const logger = new core.Logger('S3Image');
const AmplifyS3Image = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        /** The content type header used when uploading to S3 */
        this.contentType = 'binary/octet-stream';
        /** The access level of the image */
        this.level = storageTypes.AccessLevel.Public;
    }
    async watchHandler() {
        await this.load();
    }
    async componentWillLoad() {
        await this.load();
    }
    async load() {
        const { imgKey, path, body, contentType, level, track, identityId } = this;
        if (!imgKey && !path) {
            logger.debug('empty imgKey and path');
            return;
        }
        const key = imgKey || path;
        logger.debug('loading ' + key + '...');
        try {
            if (body) {
                await storageHelpers.putStorageObject(imgKey, body, level, track, contentType, logger);
            }
            this.src = await storageHelpers.getStorageObject(key, level, track, identityId, logger);
        }
        catch (err) {
            logger.debug(err);
            throw new Error(err);
        }
    }
    render() {
        return (index.h(index.Host, null, this.src && index.h("img", { src: this.src, onLoad: this.handleOnLoad, onError: this.handleOnError })));
    }
    static get watchers() { return {
        "body": ["watchHandler"]
    }; }
};
AmplifyS3Image.style = amplifyS3ImageCss;

exports.amplify_s3_image = AmplifyS3Image;
