// @aws-cdk/aws-lambda/lib/lambda-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class LambdaMetrics {
    static durationp90(dimensions) {
        return {
            namespace: 'AWS/Lambda',
            metricName: 'Duration',
            dimensions,
            statistic: 'p90',
        };
    }
    static errorsSum(dimensions) {
        return {
            namespace: 'AWS/Lambda',
            metricName: 'Errors',
            dimensions,
            statistic: 'Sum',
        };
    }
    static invocationsSum(dimensions) {
        return {
            namespace: 'AWS/Lambda',
            metricName: 'Invocations',
            dimensions,
            statistic: 'Sum',
        };
    }
    static throttlesSum(dimensions) {
        return {
            namespace: 'AWS/Lambda',
            metricName: 'Throttles',
            dimensions,
            statistic: 'Sum',
        };
    }
}
exports.LambdaMetrics = LambdaMetrics;
