// @aws-cdk/aws-stepfunctions-tasks/lib/invoke-activity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvokeActivity = void 0;
/**
 * (deprecated) A Step Functions Task to invoke an Activity worker.
 *
 * An Activity can be used directly as a Resource.
 *
 * @deprecated - use `StepFunctionsInvokeActivity`
 */
class InvokeActivity {
    /**
     * @deprecated
     */
    constructor(activity, props = {}) {
        this.activity = activity;
        this.props = props;
    }
    /**
     * (deprecated) Called when the task object is used in a workflow.
     *
     * @deprecated
     */
    bind(_task) {
        return {
            resourceArn: this.activity.activityArn,
            metricDimensions: { ActivityArn: this.activity.activityArn },
            heartbeat: this.props.heartbeat,
            // No IAM permissions necessary, execution role implicitly has Activity permissions.
            metricPrefixSingular: 'Activity',
            metricPrefixPlural: 'Activities',
        };
    }
}
exports.InvokeActivity = InvokeActivity;
