// @aws-cdk/aws-efs/lib/efs-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.EFSMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class EFSMetrics {
    static dataReadIoBytesAverage(dimensions) {
        return {
            namespace: 'AWS/EFS',
            metricName: 'DataReadIOBytes',
            dimensions,
            statistic: 'Average',
        };
    }
    static dataWriteIoBytesAverage(dimensions) {
        return {
            namespace: 'AWS/EFS',
            metricName: 'DataWriteIOBytes',
            dimensions,
            statistic: 'Average',
        };
    }
    static clientConnectionsSum(dimensions) {
        return {
            namespace: 'AWS/EFS',
            metricName: 'ClientConnections',
            dimensions,
            statistic: 'Sum',
        };
    }
    static burstCreditBalanceAverage(dimensions) {
        return {
            namespace: 'AWS/EFS',
            metricName: 'BurstCreditBalance',
            dimensions,
            statistic: 'Average',
        };
    }
    static percentIoLimitAverage(dimensions) {
        return {
            namespace: 'AWS/EFS',
            metricName: 'PercentIOLimit',
            dimensions,
            statistic: 'Average',
        };
    }
    static permittedThroughputAverage(dimensions) {
        return {
            namespace: 'AWS/EFS',
            metricName: 'PermittedThroughput',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.EFSMetrics = EFSMetrics;
