// @aws-cdk/core/lib/private/metadata-resource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataResource = void 0;
const cxapi = require("@aws-cdk/cx-api");
const region_info_1 = require("@aws-cdk/region-info");
const cfn_condition_1 = require("../cfn-condition");
const cfn_fn_1 = require("../cfn-fn");
const cfn_pseudo_1 = require("../cfn-pseudo");
const cfn_resource_1 = require("../cfn-resource");
const construct_compat_1 = require("../construct-compat");
const lazy_1 = require("../lazy");
const token_1 = require("../token");
const runtime_info_1 = require("./runtime-info");
/**
 * Construct that will render the metadata resource
 */
class MetadataResource extends construct_compat_1.Construct {
    constructor(scope, id) {
        super(scope, id);
        const metadataServiceExists = token_1.Token.isUnresolved(scope.region) || region_info_1.RegionInfo.get(scope.region).cdkMetadataResourceAvailable;
        if (metadataServiceExists) {
            const resource = new cfn_resource_1.CfnResource(this, 'Default', {
                type: 'AWS::CDK::Metadata',
                properties: {
                    Modules: lazy_1.Lazy.stringValue({ produce: () => MetadataResource.modulesProperty() }),
                },
            });
            // In case we don't actually know the region, add a condition to determine it at deploy time
            if (token_1.Token.isUnresolved(scope.region)) {
                const condition = new cfn_condition_1.CfnCondition(this, 'Condition', {
                    expression: makeCdkMetadataAvailableCondition(),
                });
                // To not cause undue template changes
                condition.overrideLogicalId('CDKMetadataAvailable');
                resource.cfnOptions.condition = condition;
            }
        }
    }
    /**
     * Clear the modules cache
     *
     * The next time the MetadataResource is rendered, it will do a lookup of the
     * modules from the NodeJS module cache again.
     *
     * Used only for unit tests.
     */
    static clearModulesCache() {
        this._modulesPropertyCache = undefined;
    }
    /**
     * Calculate the modules property
     */
    static modulesProperty() {
        if (this._modulesPropertyCache === undefined) {
            this._modulesPropertyCache = formatModules(runtime_info_1.collectRuntimeInformation());
        }
        return this._modulesPropertyCache;
    }
}
exports.MetadataResource = MetadataResource;
function makeCdkMetadataAvailableCondition() {
    return cfn_fn_1.Fn.conditionOr(...region_info_1.RegionInfo.regions
        .filter(ri => ri.cdkMetadataResourceAvailable)
        .map(ri => cfn_fn_1.Fn.conditionEquals(cfn_pseudo_1.Aws.REGION, ri.name)));
}
function formatModules(runtime) {
    const modules = new Array();
    // inject toolkit version to list of modules
    const cliVersion = process.env[cxapi.CLI_VERSION_ENV];
    if (cliVersion) {
        modules.push(`aws-cdk=${cliVersion}`);
    }
    for (const key of Object.keys(runtime.libraries).sort()) {
        modules.push(`${key}=${runtime.libraries[key]}`);
    }
    return modules.join(',');
}
