// @aws-amplify/ui-components/dist/esm/amplify-s3-image.entry.js
import { r as registerInstance, h, H as Host } from './index-39969785.js';
import { Logger } from '@aws-amplify/core';
import { A as AccessLevel } from './storage-types-f257c0f2.js';
import './constants-d1abe7de.js';
import '@aws-amplify/storage';
import { p as putStorageObject, g as getStorageObject } from './storage-helpers-1c8a0f88.js';

const amplifyS3ImageCss = ":host{height:inherit;width:inherit;--height:inherit;--width:inherit}img{height:var(--height);width:var(--width)}";

const logger = new Logger('S3Image');
const AmplifyS3Image = class {
    constructor(hostRef) {
        registerInstance(this, hostRef);
        /** The content type header used when uploading to S3 */
        this.contentType = 'binary/octet-stream';
        /** The access level of the image */
        this.level = AccessLevel.Public;
    }
    async watchHandler() {
        await this.load();
    }
    async componentWillLoad() {
        await this.load();
    }
    async load() {
        const { imgKey, path, body, contentType, level, track, identityId } = this;
        if (!imgKey && !path) {
            logger.debug('empty imgKey and path');
            return;
        }
        const key = imgKey || path;
        logger.debug('loading ' + key + '...');
        try {
            if (body) {
                await putStorageObject(imgKey, body, level, track, contentType, logger);
            }
            this.src = await getStorageObject(key, level, track, identityId, logger);
        }
        catch (err) {
            logger.debug(err);
            throw new Error(err);
        }
    }
    render() {
        return (h(Host, null, this.src && h("img", { src: this.src, onLoad: this.handleOnLoad, onError: this.handleOnError })));
    }
    static get watchers() { return {
        "body": ["watchHandler"]
    }; }
};
AmplifyS3Image.style = amplifyS3ImageCss;

export { AmplifyS3Image as amplify_s3_image };
