// @aws-cdk/aws-sns/lib/sns-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SNSMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class SNSMetrics {
    static numberOfNotificationsDeliveredAverage(dimensions) {
        return {
            namespace: 'AWS/SNS',
            metricName: 'NumberOfNotificationsDelivered',
            dimensions,
            statistic: 'Average',
        };
    }
    static numberOfNotificationsFailedAverage(dimensions) {
        return {
            namespace: 'AWS/SNS',
            metricName: 'NumberOfNotificationsFailed',
            dimensions,
            statistic: 'Average',
        };
    }
    static numberOfMessagesPublishedAverage(dimensions) {
        return {
            namespace: 'AWS/SNS',
            metricName: 'NumberOfMessagesPublished',
            dimensions,
            statistic: 'Average',
        };
    }
    static publishSizeAverage(dimensions) {
        return {
            namespace: 'AWS/SNS',
            metricName: 'PublishSize',
            dimensions,
            statistic: 'Average',
        };
    }
    static smsSuccessRateAverage(dimensions) {
        return {
            namespace: 'AWS/SNS',
            metricName: 'SMSSuccessRate',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.SNSMetrics = SNSMetrics;
