// @aws-sdk/middleware-user-agent/dist/cjs/middleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUserAgentPlugin = exports.getUserAgentMiddlewareOptions = exports.userAgentMiddleware = void 0;
const protocol_http_1 = require("@aws-sdk/protocol-http");
function userAgentMiddleware(options) {
    return (next) => (args) => {
        const { request } = args;
        if (!protocol_http_1.HttpRequest.isInstance(request))
            return next(args);
        const { headers } = request;
        const userAgentHeader = options.runtime === "node" ? "user-agent" : "x-amz-user-agent";
        if (!headers[userAgentHeader]) {
            headers[userAgentHeader] = `${options.defaultUserAgent}`;
        }
        else {
            headers[userAgentHeader] += ` ${options.defaultUserAgent}`;
        }
        if (options.customUserAgent) {
            headers[userAgentHeader] += ` ${options.customUserAgent}`;
        }
        return next({
            ...args,
            request,
        });
    };
}
exports.userAgentMiddleware = userAgentMiddleware;
exports.getUserAgentMiddlewareOptions = {
    name: "getUserAgentMiddleware",
    step: "build",
    tags: ["SET_USER_AGENT", "USER_AGENT"],
};
exports.getUserAgentPlugin = (config) => ({
    applyToStack: (clientStack) => {
        clientStack.add(userAgentMiddleware(config), exports.getUserAgentMiddlewareOptions);
    },
});
