// @aws-amplify/ui-react/lib-esm/components.js
/* eslint-disable */
/* tslint:disable */
/* auto-generated react proxies */
import { createReactComponent } from './react-component-lib';
import { defineCustomElements, applyPolyfills } from '@aws-amplify/ui-components/loader';
applyPolyfills().then(function () { return defineCustomElements(); });
export var AmplifyAmazonButton = /*@__PURE__*/ createReactComponent('amplify-amazon-button');
export var AmplifyAuthFields = /*@__PURE__*/ createReactComponent('amplify-auth-fields');
export var AmplifyAuth0Button = /*@__PURE__*/ createReactComponent('amplify-auth0-button');
export var AmplifyAuthenticator = /*@__PURE__*/ createReactComponent('amplify-authenticator');
export var AmplifyButton = /*@__PURE__*/ createReactComponent('amplify-button');
export var AmplifyChatbot = /*@__PURE__*/ createReactComponent('amplify-chatbot');
export var AmplifyCheckbox = /*@__PURE__*/ createReactComponent('amplify-checkbox');
export var AmplifyCodeField = /*@__PURE__*/ createReactComponent('amplify-code-field');
export var AmplifyConfirmSignIn = /*@__PURE__*/ createReactComponent('amplify-confirm-sign-in');
export var AmplifyConfirmSignUp = /*@__PURE__*/ createReactComponent('amplify-confirm-sign-up');
export var AmplifyContainer = /*@__PURE__*/ createReactComponent('amplify-container');
export var AmplifyCountryDialCode = /*@__PURE__*/ createReactComponent('amplify-country-dial-code');
export var AmplifyEmailField = /*@__PURE__*/ createReactComponent('amplify-email-field');
export var AmplifyFacebookButton = /*@__PURE__*/ createReactComponent('amplify-facebook-button');
export var AmplifyFederatedButtons = /*@__PURE__*/ createReactComponent('amplify-federated-buttons');
export var AmplifyFederatedSignIn = /*@__PURE__*/ createReactComponent('amplify-federated-sign-in');
export var AmplifyForgotPassword = /*@__PURE__*/ createReactComponent('amplify-forgot-password');
export var AmplifyFormField = /*@__PURE__*/ createReactComponent('amplify-form-field');
export var AmplifyFormSection = /*@__PURE__*/ createReactComponent('amplify-form-section');
export var AmplifyGoogleButton = /*@__PURE__*/ createReactComponent('amplify-google-button');
export var AmplifyGreetings = /*@__PURE__*/ createReactComponent('amplify-greetings');
export var AmplifyHint = /*@__PURE__*/ createReactComponent('amplify-hint');
export var AmplifyIcon = /*@__PURE__*/ createReactComponent('amplify-icon');
export var AmplifyIconButton = /*@__PURE__*/ createReactComponent('amplify-icon-button');
export var AmplifyInput = /*@__PURE__*/ createReactComponent('amplify-input');
export var AmplifyLabel = /*@__PURE__*/ createReactComponent('amplify-label');
export var AmplifyLink = /*@__PURE__*/ createReactComponent('amplify-link');
export var AmplifyLoadingSpinner = /*@__PURE__*/ createReactComponent('amplify-loading-spinner');
export var AmplifyNav = /*@__PURE__*/ createReactComponent('amplify-nav');
export var AmplifyOauthButton = /*@__PURE__*/ createReactComponent('amplify-oauth-button');
export var AmplifyPasswordField = /*@__PURE__*/ createReactComponent('amplify-password-field');
export var AmplifyPhoneField = /*@__PURE__*/ createReactComponent('amplify-phone-field');
export var AmplifyPhotoPicker = /*@__PURE__*/ createReactComponent('amplify-photo-picker');
export var AmplifyPicker = /*@__PURE__*/ createReactComponent('amplify-picker');
export var AmplifyRadioButton = /*@__PURE__*/ createReactComponent('amplify-radio-button');
export var AmplifyRequireNewPassword = /*@__PURE__*/ createReactComponent('amplify-require-new-password');
export var AmplifyS3Album = /*@__PURE__*/ createReactComponent('amplify-s3-album');
export var AmplifyS3Image = /*@__PURE__*/ createReactComponent('amplify-s3-image');
export var AmplifyS3ImagePicker = /*@__PURE__*/ createReactComponent('amplify-s3-image-picker');
export var AmplifyS3Text = /*@__PURE__*/ createReactComponent('amplify-s3-text');
export var AmplifyS3TextPicker = /*@__PURE__*/ createReactComponent('amplify-s3-text-picker');
export var AmplifySection = /*@__PURE__*/ createReactComponent('amplify-section');
export var AmplifySelect = /*@__PURE__*/ createReactComponent('amplify-select');
export var AmplifySelectMfaType = /*@__PURE__*/ createReactComponent('amplify-select-mfa-type');
export var AmplifySignIn = /*@__PURE__*/ createReactComponent('amplify-sign-in');
export var AmplifySignInButton = /*@__PURE__*/ createReactComponent('amplify-sign-in-button');
export var AmplifySignOut = /*@__PURE__*/ createReactComponent('amplify-sign-out');
export var AmplifySignUp = /*@__PURE__*/ createReactComponent('amplify-sign-up');
export var AmplifyStrike = /*@__PURE__*/ createReactComponent('amplify-strike');
export var AmplifyToast = /*@__PURE__*/ createReactComponent('amplify-toast');
export var AmplifyTooltip = /*@__PURE__*/ createReactComponent('amplify-tooltip');
export var AmplifyTotpSetup = /*@__PURE__*/ createReactComponent('amplify-totp-setup');
export var AmplifyUsernameField = /*@__PURE__*/ createReactComponent('amplify-username-field');
export var AmplifyVerifyContact = /*@__PURE__*/ createReactComponent('amplify-verify-contact');
