// @aws-cdk/aws-apigateway/lib/apigateway-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGatewayMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class ApiGatewayMetrics {
    static _5XxErrorSum(dimensions) {
        return {
            namespace: 'AWS/ApiGateway',
            metricName: '5XXError',
            dimensions,
            statistic: 'Sum',
        };
    }
    static latencyAverage(dimensions) {
        return {
            namespace: 'AWS/ApiGateway',
            metricName: 'Latency',
            dimensions,
            statistic: 'Average',
        };
    }
    static countSum(dimensions) {
        return {
            namespace: 'AWS/ApiGateway',
            metricName: 'Count',
            dimensions,
            statistic: 'Sum',
        };
    }
    static _4XxErrorSum(dimensions) {
        return {
            namespace: 'AWS/ApiGateway',
            metricName: '4XXError',
            dimensions,
            statistic: 'Sum',
        };
    }
    static integrationLatencyAverage(dimensions) {
        return {
            namespace: 'AWS/ApiGateway',
            metricName: 'IntegrationLatency',
            dimensions,
            statistic: 'Average',
        };
    }
    static cacheHitCountSum(dimensions) {
        return {
            namespace: 'AWS/ApiGateway',
            metricName: 'CacheHitCount',
            dimensions,
            statistic: 'Sum',
        };
    }
    static cacheMissCountSum(dimensions) {
        return {
            namespace: 'AWS/ApiGateway',
            metricName: 'CacheMissCount',
            dimensions,
            statistic: 'Sum',
        };
    }
}
exports.ApiGatewayMetrics = ApiGatewayMetrics;
